// @build-tracker/server/dist/commands/run.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Copyright (c) 2019 Paul Armstrong
 */
var path = __importStar(require("path"));
var server_1 = __importDefault(require("../server"));
exports.command = 'run';
exports.description = 'Run the Build Tracker server';
var group = 'Run';
exports.builder = function (yargs) {
    return yargs.usage("Usage: $0 " + exports.command).option('config', {
        alias: 'c',
        coerce: function (v) { return path.join(process.cwd(), v); },
        default: 'build-tracker.config.js',
        description: 'path to the build-tracker config file',
        group: group,
        normalize: true
    });
};
exports.handler = function (args) {
    var config = require(args.config);
    server_1.default(config);
};
