// @build-tracker/plugin-with-postgres/dist/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Copyright (c) 2019 Paul Armstrong
 */
var dotenv_1 = require("dotenv");
var queries_1 = __importDefault(require("./queries"));
var setup_1 = __importDefault(require("./setup"));
var pg_1 = require("pg");
dotenv_1.config();
function withPostgres(config) {
    var pgConfig = config.pg;
    var pool = new pg_1.Pool(pgConfig);
    var queries = new queries_1.default(pool);
    return __assign({}, config, { setup: setup_1.default(pool), queries: {
            build: {
                byRevision: queries.getByRevision,
                insert: queries.insert
            },
            builds: {
                byRevisions: queries.getByRevisions,
                byRevisionRange: queries.getByRevisionRange,
                byTimeRange: queries.getByTimeRange,
                recent: queries.getRecent
            }
        } });
}
exports.default = withPostgres;
